packages/server/utils/OpenAIServerManager.ts

```diff
@@ -43,107 +41,101 @@ class OpenAIServerManager {
     }
   }
 
-  // setting the minimal max_tokens limit increases the speed of the request and reduces the tokens used
-  estimateOutputTokens(reflectionsText: string[]) {
-    const avgGroupNameLength = 10 // assume an average group name length of 10 tokens
-    const totalLengthInTokens = reflectionsText.reduce(
-      (total, reflectionText) => total + Math.ceil(reflectionText.length / AVG_CHARS_PER_TOKEN),
-      0
-    )
-    const avgReflectionsPerGroup = 3 // assume each group will have around 3 reflections
-    const numGroups = Math.ceil(reflectionsText.length / avgReflectionsPerGroup)
-
-    const outputTokens = Math.round(numGroups * avgGroupNameLength + totalLengthInTokens)
+  async generateThemes(reflectionsText: string[]) {
+    if (!this.openAIApi) return null
+    const suggestedThemeCountMin = Math.floor(reflectionsText.length / 5)
+    const suggestedThemeCountMax = Math.floor(reflectionsText.length / 4)
+    const prompt = `Create a short list of common themes given the following reflections: ${reflectionsText.join(
+      ', '
+    )}. Each theme should be no longer than a few words. There should be roughly ${suggestedThemeCountMin} or ${suggestedThemeCountMax} themes. Return the themes as a comma-separated list.`
 
-    return Math.min(MAX_GPT_3_5_TOKENS, outputTokens)
+    try {
+      const response = await this.openAIApi.createChatCompletion({
+        model: 'gpt-4',
+        messages: [
+          {
+            role: 'user',
+            content: prompt
+          }
+        ],
+        temperature: 0.7,
+        top_p: 1,
+        frequency_penalty: 0,
+        presence_penalty: 0
+      })
+      const themes = (response.data.choices[0]?.message?.content?.trim() as string) ?? null
+      return themes.split(', ')
```

Comment:
is it possible that some theme would include `,` in the theme title?

Reply:
Maybe - I am requesting a comma-separated list, so it shouldn't. I haven't seen it do this after a lot of testing locally, but something to watch out for when testing internally